### PR TITLE
Add `docker.cpu.limit` and `containerd.cpu.limit` metrics based on container limits

### PR DESCRIPTION
**Pull request by vboulineau**
### What does this PR do?

Add `docker.cpu.limit` metric which reports CPU Limit in the unit as `docker.cpu.usage`. It allows to build a metric like `100 * docker.cpu.usage / docker.cpu.limit` representing the normalized CPU % per container.

Metric is based on cgroupfs `cpu.cfs_quota_us`, it works on ECS as we also fetch metrics of parent cgroup if any. It does not account for CPU Shares settings as shares are not a limit.

For `containerd`, limits are taken from container Spec, but parent cgroup limits are not supported.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: pkg/process/checks/container.go [hunk 1]
```diff
@@ -160,11 +160,11 @@ func fmtContainers(ctrList []*containers.Container, lastRates map[string]util.Co
 		containers = append(containers, &model.Container{
 			Id:          ctr.ID,
 			Type:        ctr.Type,
-			CpuLimit:    float32(ctr.CPULimit),
+			CpuLimit:    float32(ctr.Limits.CPULimit),
```
Comment (vboulineau) [inline pkg/process/checks/container.go:163]: @DataDog/processes Currently CPU Limit is computed differently whether containers have limits or not.
Let's take an example:
- We have a 4 CPUs node
- We have 1 container running without limit (containerA)
- We have 1 container running with a limit set to 2 CPUs (containerB)

Currently, the `CPULimit` value will be:
- containerA = 100%
- containerB = 200%

While, in reality, `containerA` can use up to 4 CPUs. This PR fixes this and we'll now return `400%` for `containerA`.

File: pkg/util/containers/providers/cgroup/cgroup_metrics.go [hunk 2]
```diff
@@ -24,6 +26,10 @@ import (
 // TODO: get USER_HZ from gopsutil? Needs to patch it
 const NanoToUserHZDivisor float64 = 1e9 / 100
 
+var (
+	numCPU float64 = float64(runtime.NumCPU())
```
Comment (clamoriniere) [inline pkg/util/containers/providers/cgroup/cgroup_metrics.go:30]: this value is the number of cpu seen by the agent?

In the context of running in a container? does the value may be different?

File: pkg/collector/corechecks/containers/containerd.go [hunk 3]
```diff
@@ -336,12 +345,28 @@ func parseAndSubmitMem(metricName string, sender aggregator.Sender, stat *v1.Mem
...
+
 	sender.Rate("containerd.cpu.system", float64(cpu.Usage.Kernel), "", tags)
 	sender.Rate("containerd.cpu.total", float64(cpu.Usage.Total), "", tags)
 	sender.Rate("containerd.cpu.user", float64(cpu.Usage.User), "", tags)
-	sender.Rate("containerd.cpu.throttled.periods", float64(cpu.Throttling.ThrottledPeriods), "", tags)
 
+	if cpu.Throttling != nil {
+		sender.Rate("containerd.cpu.throttled.periods", float64(cpu.Throttling.ThrottledPeriods), "", tags)
+		sender.Rate("containerd.cpu.throttled.time", float64(cpu.Throttling.ThrottledTime), "", tags)
+	}
```
Comment (clamoriniere) [inline pkg/collector/corechecks/containers/containerd.go:360]: should we return the values "0" in throttling is `nil`?

File: pkg/util/containers/providers/cgroup/cgroup_metrics.go (hunk 2 above)
Comment (vboulineau) [reply pkg/util/containers/providers/cgroup/cgroup_metrics.go:30]: Yes it's the value seen by the agent. In my knowledge there's no way to hide/modify it in a container (unlike VMs - it's probably coming from `/proc/cpuinfo`), so it's always number of CPUs on the host.

File: pkg/collector/corechecks/containers/containerd.go [hunk 4]
```diff
@@ -336,12 +345,28 @@ func parseAndSubmitMem(metricName string, sender aggregator.Sender, stat *v1.Mem
...
+
+	timeDiff := float64(currentTime.Sub(startTime).Nanoseconds()) // cpu.total is in nanoseconds
+	if timeDiff > 0 {
+		cpuLimitPct := float64(runtime.NumCPU())
+		if cpuLimits != nil && cpuLimits.Period != nil && cpuLimits.Quota != nil && *cpuLimits.Quota > 0 {
+			cpuLimitPct = float64(*cpuLimits.Quota) / float64(*cpuLimits.Period)
```
Comment (clamoriniere) [inline pkg/collector/corechecks/containers/containerd.go:366]: should we check that `*cpuLimits.Period` > 0 ?

File: pkg/collector/corechecks/containers/containerd.go (hunk 3 above)
Comment (vboulineau) [reply pkg/collector/corechecks/containers/containerd.go:360]: Previous behavior was to send nothing (incl. `containerd.cpu.usage` if `Throttling` is nil). It did not make sense to me, so I allowed to send CPU Usage anyway and kept previous behavior for `Throttling`.

About `Throttling` itself, values inside the struct are gauges (ever increasing), so if we have value once, we should always have values. So I believe, current behavior is correct.

File: pkg/collector/corechecks/containers/docker.go [hunk 5]
```diff
@@ -369,12 +360,34 @@ func (d *DockerCheck) Run() error {
...
+func (d *DockerCheck) reportCPUMetrics(cpu *cmetrics.ContainerCPUStats, limits *cmetrics.ContainerLimits, startTime int64, tags []string, sender aggregator.Sender) {
+	if cpu == nil {
+		return
+	}
+
+	sender.Rate("docker.cpu.system", float64(cpu.System), "", tags)
+	sender.Rate("docker.cpu.user", float64(cpu.User), "", tags)
+	sender.Rate("docker.cpu.usage", cpu.UsageTotal, "", tags)
+	sender.Gauge("docker.cpu.shares", float64(cpu.Shares), "", tags)
+	sender.Rate("docker.cpu.throttled", float64(cpu.NrThrottled), "", tags)
```
Comment (clamoriniere) [inline pkg/collector/corechecks/containers/docker.go:378]: Why not reporting also the `[period](https://github.com/DataDog/datadog-agent/pull/5716/files#diff-725b5ae93959fca9ffbc39b38ca40c47R109)` here

File: pkg/collector/corechecks/containers/containerd.go (hunk 4 above)
Comment (vboulineau) [reply pkg/collector/corechecks/containers/containerd.go:366]: It should be impossible to have `cpuLimits.Period == 0` with `cpuLimits.Quota > 0`, but in doubt I'll add a check on it.

**Review by clamoriniere — COMMENTED** [review]
few comments, but overall the PR looks great 👍

File: pkg/collector/corechecks/containers/docker.go (hunk 5 above)
Comment (vboulineau) [reply pkg/collector/corechecks/containers/docker.go:378]: I should remove `NrPeriod`, I used it in the first version of the PR (most accurate way to report limits), but I abandoned it as it's only available when a limit is set (and required more work on agent side).

I can add the report of `ThrottledTime` in `docker` though

File: pkg/collector/corechecks/containers/containerd.go (hunk 4 above)
Comment (vboulineau) [reply pkg/collector/corechecks/containers/containerd.go:366]: Fixed

File: pkg/collector/corechecks/containers/docker.go (hunk 5 above)
Comment (vboulineau) [reply pkg/collector/corechecks/containers/docker.go:378]: Fixed

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**